src/details/ArborX_DetailsSortUtils.hpp

```diff
@@ -65,79 +65,156 @@ namespace ArborX
-    ViewType &view)
+    ValuesType &values)
 {
-  int const n = view.extent(0);
+  static_assert(Kokkos::is_view<ValuesType>::value, "");
+  static_assert(ValuesType::rank == 1, "");
+  static_assert(std::is_same<std::decay_t<decltype(space)>,
+                             typename ValuesType::execution_space>::value,
+                "");
+
+  int const n = values.extent(0);
+
+#if defined(KOKKOS_ENABLE_CUDA)
+  auto const execution_policy = thrust::cuda::par.on(space.cuda_stream());
+#else
+  auto const execution_policy = thrust::hip::par.on(space.hip_stream());
+#endif
+
+  using ValueType = typename ValuesType::value_type;
+
+  auto values_first = thrust::device_ptr<ValueType>(values.data());
+  auto values_last = thrust::device_ptr<ValueType>(values.data() + n);
+  thrust::sort(execution_policy, values_first, values_last);
+}
+#endif
+
+template <typename ExecutionSpace, typename KeysType, typename ValuesType>
+void sortByKey(ExecutionSpace const &space, KeysType &keys, ValuesType &values)
+{
+  static_assert(Kokkos::is_view<KeysType>::value, "");
+  static_assert(Kokkos::is_view<ValuesType>::value, "");
+  static_assert(KeysType::rank == 1, "");
+  static_assert(ValuesType::rank == 1, "");
+  ARBORX_ASSERT(values.extent(0) == keys.extent(0));
+
+  int const n = keys.extent(0);
 
-  using ValueType = typename ViewType::value_type;
+  using SizeType = unsigned int;
+  using ValueType = typename KeysType::value_type;
+  using CompType = Kokkos::BinOp1D<KeysType>;
+
+  Kokkos::MinMaxScalar<ValueType> result;
+  Kokkos::MinMax<ValueType> reducer(result);
+  parallel_reduce("ArborX::Sorting::find_min_max_view",
+                  Kokkos::RangePolicy<ExecutionSpace>(space, 0, n),
+                  Kokkos::Impl::min_max_functor<KeysType>(keys), reducer);
+  if (result.min_val == result.max_val)
+    return;
+
+  Kokkos::BinSort<KeysType, CompType, typename KeysType::device_type, SizeType>
+      bin_sort(keys, CompType(n / 2, result.min_val, result.max_val), true);
+  bin_sort.create_permute_vector();
+  bin_sort.sort(keys);
+  bin_sort.sort(values);
+  // FIXME Kokkos::BinSort is currently missing overloads that take an
+  // execution space as argument
+}
+
+#if defined(KOKKOS_ENABLE_CUDA) ||                                             \
+    (defined(KOKKOS_ENABLE_HIP) && defined(ARBORX_ENABLE_ROCTHRUST))
+template <typename KeysType, typename ValuesType>
+void sortByKey(
+#if defined(KOKKOS_ENABLE_CUDA)
+    Kokkos::Cuda const &space,
+#else
+    Kokkos::Experimental::HIP const &space,
+#endif
+    KeysType &keys, ValuesType &values)
+{
+  static_assert(Kokkos::is_view<KeysType>::value, "");
+  static_assert(Kokkos::is_view<ValuesType>::value, "");
+  static_assert(KeysType::rank == 1, "");
+  static_assert(ValuesType::rank == 1, "");
   static_assert(std::is_same<std::decay_t<decltype(space)>,
-                             typename ViewType::execution_space>::value,
+                             typename ValuesType::execution_space>::value,
```

Comment:
We really want to make sure that the memory space of the given view can be accessed from the execution space, isn't it?

Reply:
Will fix it.